File: pluginsdk/host.go
package pluginsdk

import (
	"bytes"
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/url"
	"strconv"
	"strings"
	"time"

	"github.com/Tencent/WeKnora/pluginsdk/pluginapi"
)

// Host calls back into WeKnora on behalf of one call: the tenant and scopes
// are those the call's token carries. It is only valid while the token lives
// (a few minutes), so do not keep it beyond the call.
type Host struct {
	base  string
	token string
	http  *http.Client
}

// Host returns the Host API client of this call, or nil when the plugin was
// granted no Host API scopes.
func (c *Call) Host() *Host {
	if c.Context.Host == nil || c.Context.Host.URL == "" || c.Context.Host.Token == "" {
		return nil
	}
	return &Host{
		base: strings.TrimSuffix(c.Context.Host.URL, "/"), token: c.Context.Host.Token,
		http: &http.Client{Timeout: 30 * time.Second},
	}
}

func (h *Host) do(ctx context.Context, method, path string, query url.Values, body, out any) error {
	var rd io.Reader
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return err
		}
		rd = bytes.NewReader(b)
	}
	u := h.base + path
	if len(query) > 0 {
		u += "?" + query.Encode()
	}
	req, err := http.NewRequestWithContext(ctx, method, u, rd)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+h.token)
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	resp, err := h.http.Do(req)
	if err != nil {
		return pluginapi.Errorf(pluginapi.CodeUnavailable, "host api: %v", err)
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode/100 != 2 {
		var eb pluginapi.ErrorBody
		b, _ := io.ReadAll(io.LimitReader(resp.Body, 64<<10))
		if json.Unmarshal(b, &eb) == nil && eb.Error.Code != "" {
			return &eb.Error
		}
		return pluginapi.Errorf(pluginapi.CodeInternal, "host api answered HTTP %d", resp.StatusCode)
	}
	if out == nil || resp.StatusCode == http.StatusNoContent {
		return nil
	}
	return json.NewDecoder(resp.Body).Decode(out)
}

// KVGet reads key into v. ok is false when the key does not exist.
func (h *Host) KVGet(ctx context.Context, key string, v any) (ok bool, err error) {
	var e pluginapi.KVEntry
	err = h.do(ctx, http.MethodGet, pluginapi.HostKVPath, url.Values{"key": {key}}, nil, &e)
	if pe, isPE := pluginapi.AsError(err); isPE && pe.Code == pluginapi.CodeNotFound {
		return false, nil
	}
	if err != nil {
		return false, err
	}
	return true, json.Unmarshal(e.Value, v)
}

// KVPut stores v (as JSON) under key; ttl 0 keeps it until deleted.
func (h *Host) KVPut(ctx context.Context, key string, v any, ttl time.Duration) error {
	raw, err := json.Marshal(v)
	if err != nil {
		return err
	}
	in := pluginapi.KVPut{Key: key, Value: raw, TTLSeconds: int(ttl / time.Second)}
	return h.do(ctx, http.MethodPut, pluginapi.HostKVPath, nil, in, nil)
}

// KVDelete removes key; a missing key is not an error.
func (h *Host) KVDelete(ctx context.Context, key string) error {
	return h.do(ctx, http.MethodDelete, pluginapi.HostKVPath, url.Values{"key": {key}}, nil, nil)
}

// KVList returns a page of keys with prefix after the given key.
func (h *Host) KVList(ctx context.Context, prefix, after string, limit int) (*pluginapi.KVList, error) {
	q := url.Values{"prefix": {prefix}, "after": {after}}
	if limit > 0 {
		q.Set("limit", strconv.Itoa(limit))
	}
	var out pluginapi.KVList
	return &out, h.do(ctx, http.MethodGet, pluginapi.HostKVListPath, q, nil, &out)
}

// DataSources lists the workspace's data sources of the plugin's
// connectors (scope "datasources").
func (h *Host) DataSources(ctx context.Context) ([]pluginapi.DataSourceInfo, error) {
	var out pluginapi.DataSourceList
	if err := h.do(ctx, http.MethodGet, pluginapi.HostDataSourcesPath, nil, nil, &out); err != nil {
		return nil, err
	}
	return out.DataSources, nil
}

// SyncDataSource starts an incremental sync of one of them, unless one is
// already running (scope "datasources").
func (h *Host) SyncDataSource(ctx context.Context, id string) (*pluginapi.SyncStarted, error) {
	var out pluginapi.SyncStarted
	return &out, h.do(ctx, http.MethodPost, pluginapi.HostDataSourceSyncPath(url.PathEscape(id)), nil, struct{}{}, &out)
}

File: pluginsdk/python/src/weknora_plugin/host.py
"""The Host API client: how a plugin calls back into WeKnora during a call."""

from __future__ import annotations

import json
import urllib.error
import urllib.parse
import urllib.request
from datetime import timedelta
from typing import Any, Optional, Tuple, Union

from .protocol import HOST_DATA_SOURCES_PATH, HOST_KV_LIST_PATH, HOST_KV_PATH, ErrorCode, PluginError
from .types import DataSourceInfo, KVEntry, KVList, SyncStarted, from_wire


def _encode(value: Any) -> bytes:
    # UTF-8 as is, without spaces: the store counts bytes, and \uXXXX escapes
    # would make CJK text cost six bytes a character instead of three. Only
    # text UTF-8 cannot carry (lone surrogates) falls back to escapes.
    try:
        return json.dumps(value, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
    except UnicodeEncodeError:
        return json.dumps(value, separators=(",", ":")).encode()


def kv_value_size(value: Any) -> int:
    """The bytes value takes in the key-value store, as Host.kv_put sends
    it; the store refuses more than KV_MAX_VALUE_BYTES."""
    return len(_encode(value))


class Host:
    """Reaches the Host API with the short-lived token of one call. Use it
    within the call; the token expires in minutes."""

    def __init__(self, url: str, token: str, timeout: float = 30) -> None:
        self._base = url.rstrip("/")
        self._token = token
        self._timeout = timeout

    def _do(self, method: str, path: str, query: Optional[dict] = None, body: Any = None) -> Any:
        url = self._base + path
        if query:
            url += "?" + urllib.parse.urlencode(query)
        data = None
        headers = {"Authorization": "Bearer " + self._token}
        if body is not None:
            data = _encode(body)
            headers["Content-Type"] = "application/json"
        req = urllib.request.Request(url, data=data, method=method, headers=headers)
        try:
            with urllib.request.urlopen(req, timeout=self._timeout) as resp:
                raw = resp.read()
        except urllib.error.HTTPError as e:
            raw = e.read()
            try:
                err = json.loads(raw)["error"]
                if err.get("code"):
                    raise PluginError.from_wire(err) from None
            except (ValueError, KeyError, TypeError):
                pass
            raise PluginError(ErrorCode.INTERNAL, f"host api answered HTTP {e.code}") from None
        except (urllib.error.URLError, OSError) as e:
            raise PluginError(ErrorCode.UNAVAILABLE, f"host api: {e}") from None
        return json.loads(raw) if raw else None

    def kv_entry(self, key: str) -> Optional[KVEntry]:
        """Reads a key with its metadata, or None when it does not exist."""
        try:
            return from_wire(KVEntry, self._do("GET", HOST_KV_PATH, {"key": key}))
        except PluginError as e:
            if e.code == ErrorCode.NOT_FOUND:
                return None
            raise

    def kv_get(self, key: str, default: Any = None) -> Any:
        """Reads a key's value, or default when it does not exist."""
        entry = self.kv_entry(key)
        return default if entry is None else entry.value

    def kv_put(self, key: str, value: Any, ttl: Union[float, timedelta] = 0) -> None:
        """Stores a JSON value; ttl (seconds or a timedelta) 0 keeps it
        until deleted."""
        if isinstance(ttl, timedelta):
            ttl = ttl.total_seconds()
        body: dict = {"key": key, "value": value}
        if ttl:
            body["ttlSeconds"] = int(ttl)
        self._do("PUT", HOST_KV_PATH, body=body)

    def kv_delete(self, key: str) -> None:
        """Removes a key; a missing key is not an error."""
        self._do("DELETE", HOST_KV_PATH, {"key": key})

    def kv_list(self, prefix: str = "", after: str = "", limit: int = 0) -> KVList:
        """A page of keys with prefix, in key order after the given key;
        pass the result's next as after for the following page."""
        q = {"prefix": prefix, "after": after}
        if limit > 0:
            q["limit"] = str(limit)
        return from_wire(KVList, self._do("GET", HOST_KV_LIST_PATH, q))

    def kv_items(self, prefix: str = "") -> "Tuple[KVEntry, ...]":
        """Every entry with prefix, following pages."""
        out, after = [], ""
        while True:
            page = self.kv_list(prefix, after)
            out.extend(page.entries)
            if not page.next:
                return tuple(out)
            after = page.next

    def data_sources(self) -> "Tuple[DataSourceInfo, ...]":
        """The workspace's data sources of the plugin's connectors (scope
        "datasources")."""
        out = self._do("GET", HOST_DATA_SOURCES_PATH) or {}
        return tuple(from_wire(DataSourceInfo, d) for d in out.get("dataSources") or [])

    def sync_data_source(self, id: str) -> SyncStarted:
        """Starts an incremental sync of one of them, unless one is already
        running; status is "queued" or "running" (scope "datasources")."""
        path = HOST_DATA_SOURCES_PATH + "/" + urllib.parse.quote(id, safe="") + "/sync"
        return from_wire(SyncStarted, self._do("POST", path, body={}))
